### ISSUE TITLE
uv_dlopen returns wrong error

### ISSUE THREAD

**Issue by thejh** [joyent/libuv#286]
`uv_dlopen` returns wrong errors. The correct way is calling `dlerror()`. For example, I got an `ENOENT` although the real error was `/home/jann/gitty/node-nacl/build/Release/node_nacl.node: undefined symbol: ev_rt_now`.

**Comment by bnoordhuis** [comment]
Yes, known problem. See e9235a3 for a bit of background. The dl\* functions don't set or return error codes, the best you can get is a string containing the error message. We currently don't have good facilities for passing around arbitrary strings in `uv_err_t`.

**Comment by thejh** [comment]
@bnoordhuis Uh, but that commit isn't on the master branch, right? Why not?

**Comment by thejh** [comment]
At least on 32-bit-machines, this could work by having a special `uv_err_code` for string errors that means that the `sys_errno_` is a pointer, right? And by giving `sys_errno_` the size of a pointer, this should work everywhere, right?

**Comment by bnoordhuis** [comment]
> Uh, but that commit isn't on the master branch, right? Why not?

It's part of the v0.6. a.k.a. the stable branch. It hasn't been up-merged into master yet.

> At least on 32-bit-machines, this could work by having a special uv_err_code for string errors that means that the sys_errno_ is a pointer, right? And by giving sys_errno_ the size of a pointer, this should work everywhere, right?

Not quite. The pointer returned by `dlerror()` may not be persistent so libuv would need to make a copy. Who is going to free that memory and when? That's not part of the API contract right now and I don't want to bolt it on in an ad hoc fashion.

**Comment by joeferner** [comment]
Instead of returning the error string in the uv_err_t struct could it be stored in a last error string on the globals stack and then calls to uv_strerror could get the last error string.

**Comment by shigeki** [comment]
@bnoordhuis  Is there any updates on this recently? 
 I had the experience that it was very hard to investigate the error of dlopen() some time ago.  It seems that most of such troubles cannot be solved until the messages of dlerror() are shown to user.

If binary installation via npm will be deployed in the near future, they would be much more serious problems.

So, I've tried to show dlerror() message instead of "Cannot find module" implemented in the following branch so I'd like you to check it enough for you not to be an ad hoc solution.

https://github.com/shigeki/node/compare/shigeki_20120224

**Comment by bnoordhuis** [comment]
@shigeki: It's a good start. Can you submit it as a libuv PR first? I would like the API to look something like this:

``` diff
diff --git a/src/unix/dl.c b/src/unix/dl.c
index 41c244d..e87e4e0 100644
--- a/src/unix/dl.c
+++ b/src/unix/dl.c
@@ -22,6 +22,7 @@
 #include "uv.h"
 #include "internal.h"

+#include <string.h>
 #include <dlfcn.h>
 #include <errno.h>

@@ -67,3 +68,13 @@ uv_err_t uv_dlsym(uv_lib_t library, const char* name, void** ptr) {
   *ptr = (void*) address;
   return uv_ok_;
 }
+
+
+const char *uv_dlerror(uv_lib_t library) {
+  return strdup(dlerror());
+}
+
+
+void uv_dlerror_free(uv_lib_t library, const char *msg) {
+  free((void*)msg);
+}
```

**Comment by shigeki** [comment]
Thanks, Ben. I will submit a new PR soon.

**Comment by TooTallNate** [comment]
OMG +Infinity

**Comment by shigeki** [comment]
@TooTallNate Thanks a lot.

@bnoordhuis I've just submit #325. Please review it. After your review, I will submit a new PR to node.

**Comment by bnoordhuis** [comment]
Fixed in b55801f.
